### PR TITLE
Add tests for XEP-0045 Multi-User Chat (section 14)

### PR DESCRIPTION
**Pull request by guusdk**

### CODE REVIEW

File: src/main/java/org/jivesoftware/smackx/muc/MultiUserChatSecurityConsiderationsIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+            mucAsSeenByParticipant.join(nicknameParticipant);
+            mucAsSeenByParticipant.addMessageListener(listener);
+
+            // Execute system under test.
+            mucAsSeenByOwner.getConfigFormManager()
+                .enablePublicLogging()
+                .submitConfigurationForm();
+
+            // Verify result
+            assertResult(detected170Status, "Expected '" + conTwo.getUser()+ "' that is an occupant of room '" + mucAddress + "' (as '" + nicknameParticipant + "') to receive a notification with status code '170' after room owner '" + conOne.getUser() + "' (as '" + nicknameOwner + "') reconfigured the room to have public logging enabled (but no such status was received).");
```
Comment (Fishbowler) [inline src/main/java/org/jivesoftware/smackx/muc/MultiUserChatSecurityConsiderationsIntegrationTest.java:130]: Does this need to wait for the syncpoint? This looks like the assertion might race the listener and be potentially flappy.

Comment (guusdk) [reply src/main/java/org/jivesoftware/smackx/muc/MultiUserChatSecurityConsiderationsIntegrationTest.java:130]: I'm not sure if I understand what you're saying. This code registers the listener (which is set to signal the syncpoint) prior to owner flipping the configuration. The configuration change should be picked up by the listener, which as a result signals the syncpoint. Then, in line 130, `assertResult` waits for the syncpoint to be signalled (or times out).

Comment (Fishbowler) [reply src/main/java/org/jivesoftware/smackx/muc/MultiUserChatSecurityConsiderationsIntegrationTest.java:130]: The sync point is triggered by the listener on the participant receiving the message.

The owner fires off the config change and the code moves to assertion.

Isn't there a chance the assertion can fire before the server sends the resultant message to the participant? Or is there something inside Smack preventing that?

Comment (guusdk) [reply src/main/java/org/jivesoftware/smackx/muc/MultiUserChatSecurityConsiderationsIntegrationTest.java:130]: The call to `assertResult` will check if the syncpoint is signalled, and if not, will wait up to a timeout for that to happen.

if the syncpoint is signalled before `assertResult` is invoked, then the invocation will immediately progress. If the syncpoint is singalled at some time after `assertResult` is invoked, then t he invocation will progress until after the (async) signal was detected.

**Review by Fishbowler — APPROVED** [review]

**Review by Fishbowler — COMMENTED** [review]
Have you checked that the failures are definitely the ones you're expecting?

**Comment by guusdk** [comment]
Hmm... not really. I'm copying the exception here, for posterity (in case it disappears on subsequent runs):

```
INFO: MultiUserChatAdminRevokeMemberIntegrationTest.mucTestRemovalOfUserInMembersOnlyRoom (Normal) Start
Aug 29, 2024 2:08:11 PM org.igniterealtime.smack.inttest.SmackIntegrationTestFramework run
SEVERE: Unexpected abort because runTests() threw throwable
java.lang.IllegalArgumentException
	at org.jivesoftware.smackx.xdata.form.FillableForm.write(FillableForm.java:230)
	at org.jivesoftware.smackx.xdata.form.FillableForm.setAnswer(FillableForm.java:134)
	at org.jivesoftware.smackx.muc.MucConfigFormManager.submitConfigurationForm(MucConfigFormManager.java:470)
	at org.jivesoftware.smackx.muc.AbstractMultiUserChatIntegrationTest.createMembersOnlyMuc(AbstractMultiUserChatIntegrationTest.java:156)
	at org.jivesoftware.smackx.muc.MultiUserChatAdminRevokeMemberIntegrationTest.mucTestRemovalOfUserInMembersOnlyRoom(MultiUserChatAdminRevokeMemberIntegrationTest.java:317)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.lambda$runTests$0(SmackIntegrationTestFramework.java:476)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.runConcreteTest(SmackIntegrationTestFramework.java:556)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework$PreparedTest.run(SmackIntegrationTestFramework.java:764)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.runTests(SmackIntegrationTestFramework.java:544)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.run(SmackIntegrationTestFramework.java:277)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.main(SmackIntegrationTestFramework.java:115)

Aug 29, 2024 2:08:11 PM org.igniterealtime.smack.inttest.XmppConnectionManager disconnectAndCleanup
INFO: Successfully deleted all created accounts ✔
Exception in thread "main" java.lang.IllegalArgumentException
	at org.jivesoftware.smackx.xdata.form.FillableForm.write(FillableForm.java:230)
	at org.jivesoftware.smackx.xdata.form.FillableForm.setAnswer(FillableForm.java:134)
	at org.jivesoftware.smackx.muc.MucConfigFormManager.submitConfigurationForm(MucConfigFormManager.java:470)
	at org.jivesoftware.smackx.muc.AbstractMultiUserChatIntegrationTest.createMembersOnlyMuc(AbstractMultiUserChatIntegrationTest.java:156)
	at org.jivesoftware.smackx.muc.MultiUserChatAdminRevokeMemberIntegrationTest.mucTestRemovalOfUserInMembersOnlyRoom(MultiUserChatAdminRevokeMemberIntegrationTest.java:317)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.lambda$runTests$0(SmackIntegrationTestFramework.java:476)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.runConcreteTest(SmackIntegrationTestFramework.java:556)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework$PreparedTest.run(SmackIntegrationTestFramework.java:764)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.runTests(SmackIntegrationTestFramework.java:544)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.run(SmackIntegrationTestFramework.java:277)
	at org.igniterealtime.smack.inttest.SmackIntegrationTestFramework.main(SmackIntegrationTestFramework.java:115)
```

**Comment by guusdk** [comment]
The test breakage is related to [a recent change in Smack](https://github.com/igniterealtime/Smack/pull/614). A potential fix is in the works here: https://github.com/igniterealtime/Smack/pull/613

**Comment by guusdk** [comment]
Rebased

**Comment by Fishbowler** [comment]
What's your preference here? Merge and wait for Smack, or wait for Smack before merging?

**Comment by guusdk** [comment]
The Smack fix has already landed and is now being used by our code. The build now fails because of a 'regular' test failure (bug in Openfire, the system under test), not any longer because of that exception caused by the bug in Smack.

**Pull request merged**